text/0029-block-streaming.md

```diff
@@ -0,0 +1,239 @@
+These nodes can be merged/shared per operator cluster, without loss of
+generality.
+
+One alternative is to have the MobileCoin Foundation run a pub/sub service, but
+that introduces a centralized service that can become a single point of failure.
+
+The authors have found no other alternative that satisfies our constraints.
+
+The impact of not doing this is what we currently have, which wastes precious
+time doing S3 I/O, effectively using it as a message bus, for which it was
+simply not designed
+
+# Prior art
+[prior-art]: #prior-art
+
+* [Streamr Network is a decentralized pub/sub network](https://medium.com/streamrblog/building-the-missing-protocol-of-todays-internet-stack-a-decentralized-pub-sub-network-for-ad1f5972f3ca)
+* [Ethereum has Whisper](https://eth.wiki/concepts/whisper/whisper)
+* [IPFS also has a pubsub implementation](https://blog.ipfs.io/25-pubsub/)
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+We will need to experiment with different pub/sub implementations, and may
+switch implementations as needed. See [publish-subscribe-mechanisms] below.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+While this MCIP focuses on distributing blocks from consensus nodes, the same
+concepts can be useful for Fog and other subsystems.
+
+## Publish/Subscribe mechanisms
+[publish-subscribe-mechanisms]: #publish-subscribe-mechanisms
+
+This section outlines some possible implementations backing the public API.
+
+### Kafka
+
+[Apache Kafka](https://kafka.apache.org/) is a mature offering.
+
+Quoting [Confluent](https://www.confluent.io/what-is-apache-kafka):
+> Apache Kafka is a community distributed event streaming platform capable of
+> handling trillions of events a day. Initially conceived as a messaging queue,
+> Kafka is based on an abstraction of a distributed commit log. Since being
+> created and open sourced by LinkedIn in 2011, Kafka has quickly evolved from
+> messaging queue to a full-fledged event streaming platform.
+
+Pros:
+* Kafka is an established message bus, with a fair bit of established experience
+  hosting and operating instances, including many cloud-hosted options.
+  * Cloud providers include
+    [Confluent Cloud](https://www.confluent.io/confluent-cloud/),
+    [Amazon Managed Streaming](https://aws.amazon.com/msk/),
+    [Cloud Kafka](https://www.cloudkarafka.com/)
+  * Self-hosting is also a fairly popular option.
+  * [Managed Apache Kafka Vs. DIY: What’s The Difference And How To Choose?](https://keen.io/blog/managed-apache-kafka-vs-diy/)
+
+Cons:
+- Unclear how to safely expose Kafka interface across operators.
+- Having one Kafka broker per consensus node could negate the benefits of using
+  Kafka
+
+### Google Cloud Pub/Sub
+
+[Google Cloud also offers a Pub/Sub solution](https://cloud.google.com/pubsub/)
+
+Pros:
+* Horizontally scalable managed messaging offering.
+
+Cons:
+* Risks vendor lock-in.
```

Comment:
I'm assuming we're not considering [Amazon SNS](https://aws.amazon.com/sns/?sc_channel=ta&sc_campaign=modernmessagingapplicationintegration&sc_geo=namer&sc_country=us&sc_outcome=aware&whats-new-cards.sort-by=item.additionalFields.postDateTime&whats-new-cards.sort-order=desc) because it won't support the throughput we need, and it's hard to do multi-tenant?

Reply:
No, I just hadn't looked into it 🙃 Added.